daemons/execd/execd_commands.c

```diff
@@ -1326,19 +1326,28 @@ lrmd_rsc_execute_service_lib(lrmd_rsc_t * rsc, lrmd_cmd_t * cmd)
     action->cb_data = cmd;
 
     if (services_action_async(action, action_complete)) {
-        /* When services_action_async() returns TRUE, the callback might have
-         * been called -- in this case action_complete(), which might free cmd,
-         * so cmd cannot be used here.
+        /* The services library has taken responsibility for the action. It
+         * could be pending, blocked, or merged into a duplicate recurring
+         * action, in which case the action callback (action_complete())
+         * will be called when the action completes, otherwise the callback has
+         * already been called.
+         *
+         * action_complete() calls cmd_finalize() which can free cmd, so cmd
```

Comment:
> the action callback (action_complete()) will be called when the action completes, otherwise the callback has already been called.
> ...
> which can free cmd

Are we guaranteed that the services library will free cmd eventually? I ask because the comment says `action_complete()` will be called eventually and `cmd_finalize()` "can" free cmd. `action_complete()` doesn't even necessarily call `cmd_finalize()`; it can return early without calling `cmd_finalize()`.

I haven't traced everything. Sanity-checking here because I feel like you might have a good sense of the answer already.

Reply:
Yes, I went through it as carefully as possible. cmd_finalize() frees cmd except in the case of a non-cancelled recurring action, which this function adds to rsc->recurring_ops. Hence the comment at the end of stonith_action_complete() "If a recurring command is *not* rescheduled, its status must be PCMK_EXEC_CANCELLED, otherwise it will not be removed from recurring_ops by cmd_finalize()" which applies to non-stonith actions as well. AFAICT the code does follow that.